File: backend-java/src/main/java/org/egualpam/contexts/observable/order/application/domain/Order.java
package org.egualpam.contexts.observable.order.application.domain;

import org.egualpam.contexts.observable.shared.application.domain.AggregateId;
import org.egualpam.contexts.observable.shared.application.domain.AggregateRoot;

import java.time.Instant;

import static java.util.UUID.randomUUID;

public class Order extends AggregateRoot {

    private final AggregateId id;

    private Order(AggregateId id) {
        this.id = id;
    }

    @Override
    public AggregateId id() {
        return id;
    }

    public static Order create(String id) {
        var order = new Order(new AggregateId(id));

        var event = OrderCreated.create(
                randomUUID().toString(),
                Instant.now(),
                order
        );

        order.add(event);
        return order;
    }

    public static Order load(String id) {
        return new Order(new AggregateId(id));
    }
}

File: backend-kotlin/src/main/kotlin/org/egualpam/contexts/observable/order/application/domain/Order.kt
package org.egualpam.contexts.observable.order.application.domain

import org.egualpam.contexts.observable.shared.application.domain.AggregateId
import org.egualpam.contexts.observable.shared.application.domain.AggregateRoot
import java.time.Instant
import java.util.UUID.randomUUID

class Order(private val id: OrderId) : AggregateRoot() {
  override fun getId(): AggregateId = id

  companion object {
    fun create(id: String): Order {
      val order = Order(id = OrderId(id))

      val event = OrderCreated.create(
          id = randomUUID().toString(),
          occurredOn = Instant.now(),
          order = order,
      )

      order.add(event)

      return order
    }

    fun load(id: String): Order = Order(OrderId(id))
  }
}
